pages/client_login.js

```diff
@@ -0,0 +1,151 @@
+import React, { useEffect, useState, useRef } from "react";
+import Section from "components/section";
+import Page from "components/page";
+
+const redirectUrl = "https://chatterino.com/client_login";
+const twitchClientID = "g5zg0400k4vhrx2g6xi4hgveruamlv";
+const scopes = [
+  "user_subscriptions",
+  "user_blocks_edit", // deprecated, replaced with "user:manage:blocked_users"
+  "user_blocks_read", // deprecated, replaced with "user:read:blocked_users"
+  "user_follows_edit", // deprecated, soon to be removed later since we now use "user:edit:follows"
+  "channel_editor", // for /raid
+  "channel:moderate",
+  "channel:read:redemptions",
+  "chat:edit",
+  "chat:read",
+  "whispers:read",
+  "whispers:edit",
+  "channel_commercial", // for /commercial
+  "channel:edit:commercial", // in case twitch upgrades things in the future (and this scope is required)
+  "user:edit:follows", // for (un)following
+  "clips:edit", // for clip creation
+  "channel:manage:broadcast", // for creating stream markers with /marker command, and for the /settitle and /setgame commands
+  "user:read:blocked_users", // for getting list of blocked users
+  "user:manage:blocked_users", // for blocking/unblocking other users
+  "moderator:manage:automod", // for approving/denying automod messages
+];
+
+export default function ClientLogin() {
+  const hash = typeof window !== "undefined" ? window.location.hash : "";
+  const [oauthToken, setOauthToken] = useState(null);
+  const [user, setUser] = useState(null);
+  const [buttonColor, setButtonColor] = useState(
+    "bg-blue-500 hover:bg-blue-400"
+  );
+  const dataStringRef = useRef(null);
+  const [errorMessage, setErrorMessage] = useState(null);
+
+  useEffect(() => {
+    const oauthHash = location.hash.substr(1);
+    const oauthToken = oauthHash
+      .substr(oauthHash.indexOf("access_token="))
+      .split("&")[0]
+      .split("=")[1];
```

Comment:
In the case that there is no hash, I would expect this to fail. Non-existent hashes are represented as an empty string (see https://developer.mozilla.org/en-US/docs/Web/API/Location/hash). Would be good enough to check for an empty hash at the start and return early.

```suggestion
    if (location.hash === "") {
      return;
    }

    const oauthHash = location.hash.substr(1);
    const oauthToken = oauthHash
      .substr(oauthHash.indexOf("access_token="))
      .split("&")[0]
      .split("=")[1];
```